From 7e7604b34f531a83bf9a17584119faed0dffd361 Mon Sep 17 00:00:00 2001
From: Istio Automation <istio-testing-bot@google.com>
Date: Wed, 2 Feb 2022 13:48:15 -0800
Subject: validation: do not exit early on gateway TLS warnings (#37135)

This hides errors when we hit warnings - we should send both the errors
and warnings.

Co-authored-by: John Howard <howardjohn@google.com>
---
 pkg/config/validation/validation.go      | 5 +++--
 pkg/config/validation/validation_test.go | 8 ++++++++
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/pkg/config/validation/validation.go b/pkg/config/validation/validation.go
index ad7ae1a759..f79d3c4285 100644
--- a/pkg/config/validation/validation.go
+++ b/pkg/config/validation/validation.go
@@ -573,12 +573,13 @@ func validateTLSOptions(tls *networking.ServerTLSSettings) (v Validation) {
 			}
 		}
 	}
+
 	if len(invalidCiphers) > 0 {
-		return WrapWarning(fmt.Errorf("ignoring invalid cipher suites: %v", invalidCiphers.SortedList()))
+		v = appendWarningf(v, "ignoring invalid cipher suites: %v", invalidCiphers.SortedList())
 	}
 
 	if len(duplicateCiphers) > 0 {
-		return WrapWarning(fmt.Errorf("ignoring duplicate cipher suites: %v", duplicateCiphers.SortedList()))
+		v = appendWarningf(v, "ignoring duplicate cipher suites: %v", duplicateCiphers.SortedList())
 	}
 
 	if tls.Mode == networking.ServerTLSSettings_ISTIO_MUTUAL {
diff --git a/pkg/config/validation/validation_test.go b/pkg/config/validation/validation_test.go
index 4519ab6cd5..43e7c53c0b 100644
--- a/pkg/config/validation/validation_test.go
+++ b/pkg/config/validation/validation_test.go
@@ -1409,6 +1409,14 @@ func TestValidateTlsOptions(t *testing.T) {
 			},
 			"", "ECDHE-ECDSA-AES128-SHA",
 		},
+		{
+			"invalid cipher suites with invalid config",
+			&networking.ServerTLSSettings{
+				Mode:         networking.ServerTLSSettings_SIMPLE,
+				CipherSuites: []string{"not-a-cipher-suite"},
+			},
+			"requires a private key", "not-a-cipher-suite",
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-- 
2.35.3

